### PR TITLE
Add utilities for debugging LLVM during sysimg build

### PR DESCRIPTION
**Pull request by Keno**
This adds an extra command line option to dump the unoptimized
rather than the optimized representation of the LLVM IR. This
is very useful when debugging new LLVM passes (which we're now
increasingly making use of) or when profiling LLVM, to split
out JIT and sysimg optimization.

### CODE REVIEW

File: src/jitlayers.cpp [hunk 1]
```diff
@@ -1240,18 +1240,46 @@ void jl_dump_native(const char *bc_fname, const char *obj_fname, const char *sys
...
+        std::string err;
+        unopt_bc_OS.reset(new raw_fd_ostream(unopt_bc_fname, err, raw_fd_ostream::F_Binary));
+#endif
+        if (!err.empty())
+            jl_safe_printf("%s\n", err.c_str());
+        else {
+#if JL_LLVM_VERSION < 30700
+            unopt_bc_FOS.reset(new formatted_raw_ostream(*unopt_bc_OS.get()));
+#endif
+            PM.add(createBitcodeWriterPass(*unopt_bc_FOS.get()));     // Unroll small loops
```
Comment (vtjnash) [inline src/jitlayers.cpp]: I hope it doesn't unroll loops

File: src/jitlayers.cpp [hunk 2]
```diff
@@ -1240,18 +1240,46 @@ void jl_dump_native(const char *bc_fname, const char *obj_fname, const char *sys
...
+
+    if (unopt_bc_fname) {
+#if JL_LLVM_VERSION >= 30500
+        // call output handler directly to avoid special case handling of `-` filename
+        int FD;
+        std::error_code EC = sys::fs::openFileForWrite(unopt_bc_fname, FD, sys::fs::F_None);
+        unopt_bc_FOS.reset(new raw_fd_ostream(FD, true));
+        std::string err;
+        if (EC)
+            err = "ERROR: failed to open --output-bc file '" + std::string(unopt_bc_fname) + "': " + EC.message();
```
Comment (vtjnash) [inline src/jitlayers.cpp]: wrong argument name

File: src/jitlayers.cpp (hunk 1 above)
Comment (Keno) [inline src/jitlayers.cpp]: Ah yeah, that's been there forever. Guess I might as well fix that while I'm here.

File: src/codegen.cpp [hunk 3]
```diff
@@ -7472,11 +7472,11 @@ extern "C" void jl_init_codegen(void)
     ;
     Triple TheTriple(sys::getProcessTriple());
 #if defined(FORCE_ELF)
-#if JL_LLVM_VERSION >= 30500
-    TheTriple.setObjectFormat(Triple::ELF);
-#else
-    TheTriple.setEnvironment(Triple::ELF);
-#endif
+    #if JL_LLVM_VERSION >= 30500
```
Comment (tkelman) [inline src/codegen.cpp]: made more sense with the previous alignment, we usually left-align almost all preprocessor directives - nesting can be made clearer with comments on the `#endif` lines marking which they're closing

Comment (Keno) [inline src/codegen.cpp]: Fine with me. These conditionals will go away very soon anyway.

**Pull request merged**